Cross Site  - Gossip Router tunnel support

pruivo opened the pull request:
Adds support to use JGroups' TUNNEL protocol with GossipRouter pods.
It uses the JGroups JAR inside the server image to avoid depending on another image.

PR is open to get feedback sooner than later. No rush to integrate yet.

TODO:

- [ ] Create an entry point in the server image (avoid doing "java -jar ...")
- [ ] Handle server image upgrades (?)

pruivo commented:
depends on 
* https://github.com/infinispan/infinispan-operator/pull/1238
* https://github.com/infinispan/infinispan-operator/pull/1252

dmvolod commented on controllers/infinispan_controller.go:
```diff
@@ -180,7 +180,7 @@ func (r *InfinispanReconciler) SetupWithManager(mgr ctrl.Manager) error {
 // +kubebuilder:rbac:groups=core;events.k8s.io,resources=events,verbs=create;patch
 
-// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=create;get;list;watch
```
Seems to delete rights is missing

pruivo replied:
right! added `delete` and `update` permissions.

dmvolod commented on controllers/infinispan_controller.go:
```diff
@@ -290,6 +290,32 @@ func (reconciler *InfinispanReconciler) Reconcile(ctx context.Context, ctrlReque
 	}
 
+	if infinispan.HasSites() {
+		tunnel := &appsv1.Deployment{}
```
Whole get/create/update cycle could be replaced with controllerutil.CreateOrUpdate function

dmvolod commented on controllers/infinispan_service_controller.go:
```diff
@@ -466,8 +465,7 @@ func computeServiceExternal(ispn *ispnv1.Infinispan) *corev1.Service {
 // computeSiteService compute the XSite service
 func computeSiteService(ispn *ispnv1.Infinispan) *corev1.Service {
-	lsPodSelector := PodLabels(ispn.Name)
-	lsPodSelector[consts.CoordinatorPodLabel] = strconv.FormatBool(true)
```
@pruivo coordinator pod labeling seems to be removed also

pruivo replied:
it is not required for cross-site.

dmvolod replied:
I talked about to remove this logic. Do we really need it after GossipRouter Implementation?
https://github.com/infinispan/infinispan-operator/blob/6381d8b0fd4aca538fab2c19c97a40d6374a19b9/controllers/xsite.go#L17-L26

pruivo replied:
right! I've removed it.

dmvolod commented:
Also we need to watch Deployment resources to catch unpredictable resource remove/delete/update

pruivo commented:
> Also we need to watch Deployment resources to catch unpredictable resource remove/delete/update

I wasn't aware of that. Is there any example in the code?

dmvolod commented:
> I wasn't aware of that. Is there any example in the code?

You should add `appsv1.Deployment` to the `secondaryResourceTypes` list
https://github.com/infinispan/infinispan-operator/blob/6381d8b0fd4aca538fab2c19c97a40d6374a19b9/controllers/infinispan_controller.go#L126

dmvolod commented on controllers/infinispan_controller.go:
```diff
@@ -290,6 +290,32 @@ func (reconciler *InfinispanReconciler) Reconcile(ctx context.Context, ctrlReque
+				return reconcile.Result{}, err
+			}
+			reqLogger.Info("Creating a new Deployment", "Deployment.Name", tunnel.Name)
+			err = r.Client.Create(context.TODO(), tunnel)
```
@pruivo another potential issue, if something happens in the cluster, i.e. no resources or anything else, customers never know that Deployment not created correctly. For cluster Pods controller eventually request data from the Pod status and log it into the Infinispan CR status. Seems to need to implement similar logic for the GossipRouter Deployment

dmvolod commented on controllers/xsite.go:
```diff
@@ -5,25 +5,21 @@ import (
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 )
 
 func (r *infinispanRequest) applyLabelsToCoordinatorsPod(podList *corev1.PodList, siteLocations []string, cluster ispn.ClusterInterface) (*ispnv1.InfinispanCondition, error) {
```
This function should be named related to the task it does now, like 'getCrossSiteViewCondition'

pruivo commented:
updated the code to use `controllerutil.CreateOrUpdate()`.

Also, just crossed my mind, I should add some readiness and liveness  probes to GossipRouter pod.

pruivo commented:
code updated. 
Added probes
Added condition "GossipRouterReady" example
```yaml
Status:
  Conditions:
    Status:   True
    Type:     PreliminaryChecksPassed
    Status:   True
    Type:     GossipRouterReady
    Message:  View: test-cross-site-view-internal-xsite2-0-63738
    Status:   True
    Type:     WellFormed
    Message:  Cross-Site view: xsite1,xsite2
    Status:   True
    Type:     CrossSiteViewFormed
  ```

dmvolod commented on controllers/infinispan_controller.go:
```diff
@@ -525,7 +556,7 @@ func (reconciler *InfinispanReconciler) Reconcile(ctx context.Context, ctrlReque
 
 	// If x-site enable configure the coordinator pods to be selected by the x-site service
```
comment also need to changed

pruivo commented:
updated.
Added: shutting down the GossipRouter pod (set Deployment.Spec.Replicas=0) if Infinispan.Spec.Replicas == 0

dmvolod commented on controllers/xsite.go:
```diff
@@ -43,18 +39,64 @@ func (r *infinispanRequest) applyLabelsToCoordinatorsPod(podList *corev1.PodList
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      m.ObjectMeta.Name,
```
`ObjectMeta` can be omitted

dmvolod commented on controllers/infinispan_controller.go:
```diff
@@ -290,6 +290,37 @@ func (reconciler *InfinispanReconciler) Reconcile(ctx context.Context, ctrlReque
+			reqLogger.Error(err, "Failed to configure Cross-Site Deployment")
+			return reconcile.Result{}, err
+		}
+		reqLogger.Info("Cross-Site Deployment created")
```
@pruivo you will always print  "Cross-Site Deployment created" on each reconcile loop whenever it created or updated
second parameter of the `controllerutil.CreateOrUpdate` returns performed operation

dmvolod commented on controllers/infinispan_controller.go:
```diff
@@ -290,6 +290,37 @@ func (reconciler *InfinispanReconciler) Reconcile(ctx context.Context, ctrlReque
+		reqLogger.Info("Creating the Cross-Site Deployment (Gossip Router)")
+		tunnel := r.GetGossipRouterDeployment(infinispan)
+		_, err = controllerutil.CreateOrUpdate(r.ctx, r.Client, tunnel, func() error {
+			return controllerutil.SetControllerReference(r.infinispan, tunnel, r.scheme)
```
You just update controller reference in the mutate function, not required Deployment parts.
Please have a look how others controllers (service/config, etc.) performing updates for mutating fields

pruivo replied:
I have no idea what to do TBH. 
The service seems to copy every field manually (which seems error-prone to me). Configs, secrets, batch just invoke `SetControllerReference()`.

dmvolod replied:
The code should like this
```
		tunnelDeployment := &appsv1.Deployment{
			ObjectMeta: metav1.ObjectMeta{
				Name:      infinispan.Name,
				Namespace: infinispan.Namespace,
			},
		}
		result, err := controllerutil.CreateOrUpdate(r.ctx, r.Client, tunnelDeployment, func() error {
			tunnel := r.GetGossipRouterDeployment(infinispan)
			tunnelDeployment.Spec = tunnel.Spec
			tunnelDeployment.Labels = tunnel.Labels
			if tunnelDeployment.CreationTimestamp.IsZero() {
				reqLogger.Info("Creating the Cross-Site Deployment (Gossip Router)")
				return controllerutil.SetControllerReference(r.infinispan, tunnelDeployment, r.scheme)
			}
			return nil
		})
		if err != nil {
			if errors.IsConflict(err) {
				return reconcile.Result{Requeue: true}, nil
			} else {
				reqLogger.Error(err, "Failed to configure Cross-Site Deployment")
				return reconcile.Result{}, err
			}
		}
		if result == controllerutil.OperationResultCreated {
			reqLogger.Info("Cross-Site Deployment created")
		}
```

dmvolod commented on controllers/xsite.go:
```diff
@@ -43,18 +39,64 @@ func (r *infinispanRequest) applyLabelsToCoordinatorsPod(podList *corev1.PodList
+
+	deployment := &appsv1.Deployment{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
```
```suggestion
			APIVersion: "apps/v1",
```

dmvolod commented on controllers/xsite.go:
```diff
@@ -43,18 +39,64 @@ func (r *infinispanRequest) applyLabelsToCoordinatorsPod(podList *corev1.PodList
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: m.ObjectMeta.Namespace,
```
```suggestion
			Namespace: m.Namespace,
```

dmvolod commented on controllers/xsite.go:
```diff
@@ -43,18 +39,64 @@ func (r *infinispanRequest) applyLabelsToCoordinatorsPod(podList *corev1.PodList
+		replicas = 0
+	}
+
+	deployment := &appsv1.Deployment{
```
@pruivo what do you think about `gossiprouter` sidecar, not separate deployment? Sidecar shares network/storage with the Infinispan container, but number of the sidecars will be the same as cluster size.

pruivo replied:
There is no advantage of having more than one GossipRouter. I would have to tag a single pod to be used with the exposed service and that pod will be hit with all the traffic from both sites (and it will negate the multi site master advantage).
If that's ok for you, I would rather a the gossip router in its own pod.

dmvolod replied:
Yes, sure. The sidecar additional value is a tightly coupling within the server Pods. If it's not so important, we could live it as is.

dmvolod commented on controllers/infinispan_controller.go:
```diff
@@ -290,6 +290,37 @@ func (reconciler *InfinispanReconciler) Reconcile(ctx context.Context, ctrlReque
 		}
 	}
 
+	if infinispan.HasSites() {
```
If someone remove `sites` section from the CR, Deployment will keep forever. Seems we need validate it and cleanup if needed or set `sites` section as immutable in the webhook controller (not implemented now)

pruivo replied:
where can it be set immutable? 

Removing the deployment wouldn't help since JGroups and Infinispan are still configured to use cross-site replication. I guess it would be easier to set the field immutable than restarting and reconfiguring the cluster. 
Do you agree?

dmvolod replied:
Seem to if someone remove `sites` structure, configuration ConfgiMap will be changed and StatefulSet will be rollout to the new config without cross-site setup. Is'n it?

pruivo replied:
I'll test this locally to see what the operator does and acting accordingly. thanks!

pruivo replied:
@dmvolod you are right! 
I've updated the code to remove the `-site` `service` and the `deployment`.

dmvolod commented on controllers/xsite.go:
```diff
@@ -43,18 +39,64 @@ func (r *infinispanRequest) applyLabelsToCoordinatorsPod(podList *corev1.PodList
+			Kind:       "Deployment",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
```
Probably we need  a specific name for GossipRouter deployment, not the same as cluster name, i.e. with `tunnel` prefix or postfix

pruivo replied:
name is `name := m.GetGossipRouterDeploymentName()`
`-tunnel` is added to the cluster name.

pruivo replied:
actually, I don't need the variable `name`. I'll remove it to avoid confusion

dmvolod commented on controllers/infinispan_controller.go:
```diff
@@ -290,6 +290,55 @@ func (reconciler *InfinispanReconciler) Reconcile(ctx context.Context, ctrlReque
+		reqLogger.Info("Creating the Cross-Site Deployment (Gossip Router)")
+		tunnelDeployment := &appsv1.Deployment{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      infinispan.Name,
```
```suggestion
				Name:      infinispan.GetGossipRouterDeploymentName(),
```

ryanemerson commented:
If you're happy with the changes @dmvolod I will merge

ryanemerson commented:
Thanks  @pruivo

ryanemerson commented:
> Also Rolling Upgrade PR need to be adopted for GossipRouter deployment if it will be supported

Good point @dmvolod. FYI @gustavonalle
